apiserver/billing/odoostorage/odoo/odoo8/odoo8.go

```diff
@@ -0,0 +1,168 @@
+			return client.Open(ctx, odooURL, client.ClientOptions{UseDebugLogger: debugTransport})
+		},
+	}
+}
+
+type oodo8Storage struct {
+	sessionCreator func(ctx context.Context) (client.QueryExecutor, error)
+}
+
+func (s *oodo8Storage) Get(ctx context.Context, name string) (*billingv1.BillingEntity, error) {
+	id, err := k8sIDToOdooID(name)
+	if err != nil {
+		return nil, err
+	}
+
+	session, err := s.sessionCreator(ctx)
+	if err != nil {
+		return nil, err
+	}
+	o := model.NewOdoo(session)
+
+	accountingContact, err := o.FetchPartnerByID(ctx, id, roleAccountFilter, activeFilter)
+	if err != nil {
+		return nil, fmt.Errorf("fetching accounting contact by ID: %w", err)
+	}
+
+	if !accountingContact.Parent.Valid {
+		return nil, fmt.Errorf("accounting contact %d has no parent", id)
+	}
+
+	company, err := o.FetchPartnerByID(ctx, accountingContact.Parent.ID, activeFilter)
+	if err != nil {
+		return nil, fmt.Errorf("fetching parent %d of accounting contact %d failed: %w", accountingContact.Parent.ID, id, err)
+	}
+
+	be := mapPartnersToBillingEntity(company, accountingContact)
+	return &be, nil
+}
+
+func (s *oodo8Storage) List(ctx context.Context) ([]billingv1.BillingEntity, error) {
+	l := klog.FromContext(ctx)
+
+	session, err := s.sessionCreator(ctx)
+	if err != nil {
+		return nil, err
+	}
+	o := model.NewOdoo(session)
+
+	accPartners, err := o.SearchPartners(ctx, []client.Filter{
+		roleAccountFilter,
+		activeFilter,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	companyIDs := make([]int, 0, len(accPartners))
+	for _, p := range accPartners {
+		if !p.Parent.Valid {
+			l.Info("role account has no parent", "id", p.ID)
+			continue
+		}
+		companyIDs = append(companyIDs, p.Parent.ID)
+	}
+
+	companies, err := o.SearchPartners(ctx, []client.Filter{
+		activeFilter,
+		[]any{"id", "in", companyIDs},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	companySet := make(map[int]model.Partner, len(companies))
+	for _, p := range companies {
+		companySet[p.ID] = p
+	}
+
+	bes := make([]billingv1.BillingEntity, 0, len(accPartners))
+	for _, p := range accPartners {
+		if !p.Parent.Valid {
+			continue
+		}
+		mp, ok := companySet[p.Parent.ID]
+		if !ok {
+			l.Info("could not load parent partner (maybe no longer active?)", "parent_id", p.Parent.ID, "id", p.ID)
+			continue
+		}
+		bes = append(bes, mapPartnersToBillingEntity(mp, p))
+	}
+
+	return bes, nil
+}
+
+func (s *oodo8Storage) Create(ctx context.Context, be *billingv1.BillingEntity) error {
+	return errors.New("not implemented")
+}
+
+func (s *oodo8Storage) Update(ctx context.Context, be *billingv1.BillingEntity) error {
+	return errors.New("not implemented")
+}
+
+func k8sIDToOdooID(id string) (int, error) {
+	if !strings.HasPrefix(id, "be-") {
+		return 0, fmt.Errorf("invalid ID, missing prefix: %s", id)
+	}
+
+	return strconv.Atoi(id[3:])
+}
+
+func odooIDToK8sID(id int) string {
+	return fmt.Sprintf("be-%d", id)
+}
```

Comment:
I'm just curious, why the `be-` prefix?

Reply:
IIRC, Kubernetes doesn't like resource names which start with a digit

Reply:
Yes, it is not strictly necessary but "expected" for most resources. Also pitfalls with yaml serialisation and co.